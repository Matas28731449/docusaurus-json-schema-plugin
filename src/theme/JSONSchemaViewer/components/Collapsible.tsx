import React, {
  ReactNode,
  DetailedHTMLProps,
  DetailsHTMLAttributes,
} from "react"

// For collapse component
function Collapsible(props: {
  summary: ReactNode
  children: ReactNode
  detailsProps?: DetailedHTMLProps<
    DetailsHTMLAttributes<HTMLDetailsElement>,
    HTMLDetailsElement
  >
}): JSX.Element {
  const { summary, children, detailsProps } = props

  return (
    <details {...detailsProps}>
      <summary>{summary}</summary>
      {children}
    </details>
  )
}

export default Collapsible
