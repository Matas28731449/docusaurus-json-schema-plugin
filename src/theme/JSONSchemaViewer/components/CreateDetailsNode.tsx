import React from "react"
import Translate from "@docusaurus/Translate"

import { Collapsible } from "./index"
import { getQualifierMessages } from "../utils"

import type { JSONSchema7 } from "json-schema"

function createDetailsNode(
  name: string,
  schemaName: string,
  schema: JSONSchema7,
  required: boolean | string[]
): JSX.Element {
  // TODO : return create("SchemaItem", { at the top
  // TODO : createNodes(schema) at bottom

  const isRequired = Array.isArray(required)
    ? required.includes(name)
    : required

  let qualifierMessages = getQualifierMessages(schema)

  return (
    <Collapsible
      summary={
        <div>
          <strong>{name}</strong>
          <span style={{ opacity: "0.6" }}>{schemaName}</span>
          {isRequired && (
            <strong
              style={{
                fontSize: "var(--ifm-code-font-size)",
                color: "var(--ifm-color-danger)",
              }}
            >
              <Translate
                values={{
                  id: "json-schema.keywords.required",
                }}
              >
                {"required"}
              </Translate>
            </strong>
          )}
        </div>
      }
      children={
        <div style={{ marginLeft: "1rem" }}>
          {qualifierMessages !== undefined && (
            <div style={{ marginTop: ".5rem", marginBottom: ".5rem" }}>
              {qualifierMessages}
            </div>
          )}
          {schema?.description !== undefined && (
            <div style={{ marginTop: ".5rem", marginBottom: ".5rem" }}>
              {schema?.description}
            </div>
          )}
        </div>
      }
    />
  )
}

export default createDetailsNode
