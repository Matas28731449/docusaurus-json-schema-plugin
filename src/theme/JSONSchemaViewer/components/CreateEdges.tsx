import React from "react"
import { generateFriendlyName } from "../utils/index"
import { CreateDetailsNode } from "./index"

import type { JSONSchema7Definition } from "json-schema"

// Creates the edges or "leaves" of a schema tree. Edges can branch into sub-nodes with createDetails().
type EdgeProps = {
  name: string
  schema: JSONSchema7Definition
  required: string[] | boolean
}

// TODO
function createEdges({ name, schema, required }: EdgeProps): JSX.Element {
  const schemaName = generateFriendlyName(schema)

  // TODO review that later
  if (typeof schema === "boolean") {
    return <></>
  }

  // Most of the time, createDetailsNode do the job recursively
  return (
    <CreateDetailsNode
      name={name}
      schemaName={schemaName}
      required={required}
      schema={schema}
    />
  )

  /* TODO check that part later
  //primitives and array of non-objects
  return (
    <SchemaItem
      collapsible={false}
      name={name}
      schema={schema}
      schemaName={schemaName}
    />
  )*/
}

export default createEdges
