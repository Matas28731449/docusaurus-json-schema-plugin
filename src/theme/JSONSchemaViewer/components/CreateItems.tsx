import React from "react"
import {
  RenderAnyOneOf,
  CreateNodes,
  CreateProperties,
  CreateEdges,
  CreateAdditionalProperties,
} from "./index"

import type { JSONSchema7 } from "json-schema"
import type { WithRequired } from "./index"

type Props = {
  [x: string]: any
  schema: WithRequired<JSONSchema7, "items">
}

function createItems(props: Props): JSX.Element {
  const { schema } = props
  if (Array.isArray(schema.items)) {
    let entries = Object.entries(
      schema.items.filter((sub) => typeof sub !== "boolean") as JSONSchema7[]
    )

    return (
      <>
        {entries.map(([key, val]) => (
          <CreateEdges
            name={key}
            schema={val}
            required={
              Array.isArray(val?.required) ? val.required.includes(key) : false
            }
            key={key}
          />
        ))}
      </>
    )
  } else {
    if (typeof schema.items === "boolean" || schema.items === undefined) {
      return <></>
    }

    let typedItems = schema.items

    if (typedItems.properties !== undefined) {
      return (
        <CreateProperties
          schema={typedItems as WithRequired<JSONSchema7, "properties">}
        />
      )
    }

    if (typedItems.additionalProperties !== undefined) {
      return (
        <CreateAdditionalProperties
          schema={
            typedItems as WithRequired<JSONSchema7, "additionalProperties">
          }
        />
      )
    }

    if (typedItems.oneOf !== undefined || typedItems.anyOf !== undefined) {
      return (
        <RenderAnyOneOf
          schema={
            typedItems as
              | WithRequired<JSONSchema7, "oneOf">
              | WithRequired<JSONSchema7, "anyOf">
          }
        />
      )
    }

    // Let the createNodes do the rest
    return <CreateNodes schema={typedItems} />
  }
}

export default createItems
