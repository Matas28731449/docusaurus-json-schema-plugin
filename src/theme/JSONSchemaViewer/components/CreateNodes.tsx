import React from "react"

import { CreateTypes } from "./index"

import {
  SchemaComposition,
  SchemaConditional,
} from "../JSONSchemaElements/index"

import { CreateValidOrInvalid } from "./index"

import { isSchemaComposition, isSchemaConditional } from "../utils/index"

import type { JSONSchema } from "../types"

type Props = {
  [x: string]: any
  schema: JSONSchema
}

// Entry point
export default function CreateNodes(props: Props): JSX.Element {
  const { schema } = props

  if (typeof schema === "boolean") {
    return <CreateValidOrInvalid schema={schema} />
  }

  // Type Checks
  const isComposition = isSchemaComposition(schema)
  const isConditional = isSchemaConditional(schema)

  return (
    <>
      {/* Handle standard types */}
      <CreateTypes schema={schema} />
      {/* handle anyOf / allOf / oneOf / not  */}
      {isComposition && <SchemaComposition schema={schema} />}
      {/* Conditional part of the schema */}
      {isConditional && <SchemaConditional schema={schema} />}
    </>
  )
}
