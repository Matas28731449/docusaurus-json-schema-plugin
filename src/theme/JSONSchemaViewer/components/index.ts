// To avoid circular dependancies issue, let's have a single place for exports and imports

// Collapsible componenent
export { default as Collapsible } from "./Collapsible"

// CreateDetailsNode component
export { default as CreateDetailsNode } from "./CreateDetailsNode"

// CreateEdges component
export { default as CreateEdges } from "./CreateEdges"

// CreateProperties component
export { default as CreateProperties } from "./CreateProperties"

// RenderAnyOneOf component
export { default as RenderAnyOneOf } from "./RenderAnyOneOf"

// https://stackoverflow.com/a/69328045/6149867
export type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
