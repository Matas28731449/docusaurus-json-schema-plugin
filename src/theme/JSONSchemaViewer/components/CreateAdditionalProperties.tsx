import React from "react"
import Translate from "@docusaurus/Translate"

import { getQualifierMessages } from "../utils/index"
import { CreateNodes } from "../components/index"

import type {
  JSONSchema7,
  JSONSchema7Definition,
  JSONSchema7TypeName,
} from "json-schema"

import type { WithRequired } from "./index"

type Props = {
  schema: WithRequired<JSONSchema7, "additionalProperties">
  [x: string]: any
}

function createAdditionalProperties(props: Props): JSX.Element {
  const { schema } = props
  let typedSchema = schema.additionalProperties as JSONSchema7Definition

  // don't want to display something in boolean cases, at least from now ...
  if (typeof typedSchema === "boolean") {
    return <></>
  }

  let types = (
    Array.isArray(typedSchema?.type) ? typedSchema?.type : [typedSchema?.type]
  ).filter((s) => s !== undefined) as JSONSchema7TypeName[]

  // Usually, we have only "type" in the payload : https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties

  if (types.length > 0) {
    // Most of the time, only one entry but prefer to be safe that sorry ...
    let typesAsString = types.join(" OR ")
    let qualifierMessages = getQualifierMessages(typedSchema)

    return (
      <li>
        <div>
          <code>
            <Translate
              values={{
                id: "json-schema.labels.additionalProperties",
              }}
            >
              {"property name*"}
            </Translate>
          </code>
          <span style={{ opacity: "0.6" }}>{` ${typesAsString}`}</span>
          {typedSchema?.format !== undefined && (
            <span style={{ opacity: "0.6" }}>{` ${typedSchema.format}`}</span>
          )}
          {qualifierMessages !== undefined && (
            <div style={{ marginTop: "var(--ifm-table-cell-padding)" }}>
              {qualifierMessages}
            </div>
          )}
        </div>
      </li>
    )
  } else {
    // Well well, at this point we could have anything so let createNodes do the job
    return <CreateNodes schema={schema} />
  }
}

export default createAdditionalProperties
