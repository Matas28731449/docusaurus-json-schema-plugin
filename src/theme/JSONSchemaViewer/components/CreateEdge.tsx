import React, { ReactNode } from "react"

import { SchemaItem } from "@theme/JSONSchemaViewer/components"

import type { JSX } from "react"
import type { JSONSchema } from "@theme/JSONSchemaViewer/types"

// Creates the edge or "leave" of a schema tree. Edge can branch into sub-nodes.
type EdgeProps = {
  // Name of the attribute
  name: ReactNode
  // the schema of this attribute
  schema: JSONSchema
  // Is property is required or not
  required: boolean
  //
  onInsert?: (jsonPointer: string) => void
}

export default function CreateEdge({
  name,
  schema,
  required,
  onInsert,
}: EdgeProps): JSX.Element {
  return <SchemaItem name={name} schema={schema} required={required} onInsert={onInsert} />
}
