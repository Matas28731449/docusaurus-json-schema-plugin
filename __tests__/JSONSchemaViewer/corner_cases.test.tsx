import React from "react"

// For typings autocomplete whatever your IDE
import { expect, test, describe } from "@jest/globals"

import { create, act } from "react-test-renderer"

// Type to prevent creating invalid mocks
import type {
  JSONSchema,
  JSONSchemaNS,
} from "../../src/theme/JSONSchemaViewer/types"

// Type for react-test-renderer
import type { ReactTestRenderer } from "react-test-renderer"

// Components
import { CreateNodes } from "../../src/theme/JSONSchemaViewer/components/index"

import {
  CreateArray,
  CreateObject,
} from "../../src/theme/JSONSchemaViewer/JSONSchemaElements/index"

// Annoying test cases, just for coverage stories ...
const testcases: [
  string,
  (props: { schema: any; [x: string]: any }) => JSX.Element,
  JSONSchema
][] = [
  ["CreateNodes - boolean schema (false)", CreateNodes, false],
  ["CreateNodes - boolean schema (true)", CreateNodes, true],
  [
    "CreateArray - sub properties as boolean schema",
    CreateArray,
    {
      items: false,
      contains: false,
      prefixItems: false,
    } as JSONSchemaNS.Array,
  ],
  [
    "CreateObject - sub properties as boolean schema",
    CreateObject,
    {
      additionalProperties: false,
    } as JSONSchemaNS.Object,
  ],
  [
    "CreateObject - missing pattern in propertyNames",
    CreateObject,
    {
      propertyNames: {
        description: "I forgot the pattern key",
      },
    } as JSONSchemaNS.Object,
  ],
]

describe("JSONSchemaViewer - corner cases", () => {
  test.each(testcases)("%s", async (_title, Component, fakeSchema) => {
    // render the component
    let root: ReactTestRenderer | undefined

    await act(async () => {
      root = create(<Component schema={fakeSchema} />)
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })
})
