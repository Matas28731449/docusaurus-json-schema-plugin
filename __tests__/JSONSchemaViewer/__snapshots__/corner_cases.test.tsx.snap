// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - corner cases CreateArray - sub properties as boolean schema 1`] = `
[
  <strong>
    type
  </strong>,
  "Â :Â ",
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>,
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <strong>
            none
          </strong>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <strong>
              none
            </strong>
            <div
              style={
                {
                  "marginTop": "var(--ifm-table-cell-padding)",
                }
              }
            >
              Always invalid
            </div>
          </div>
        </div>
      </details>
    </li>
  </ul>,
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <strong>
            none
          </strong>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <strong>
              none
            </strong>
            <div
              style={
                {
                  "marginTop": "var(--ifm-table-cell-padding)",
                }
              }
            >
              Always invalid
            </div>
          </div>
        </div>
      </details>
    </li>
  </ul>,
  <li>
    <details
      open={false}
    >
      <summary>
        <code>
          items[..., x, ...]
        </code>
        Â 
        <strong>
          none
        </strong>
      </summary>
      <div>
        <div>
          <strong>
            type
          </strong>
          Â :Â 
          <strong>
            none
          </strong>
          <div
            style={
              {
                "marginTop": "var(--ifm-table-cell-padding)",
              }
            }
          >
            Always invalid
          </div>
        </div>
      </div>
    </details>
  </li>,
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      ðŸš¨Â 
      No extra item(s) are authorized in this array
    </div>
  </div>,
]
`;

exports[`JSONSchemaViewer - corner cases CreateNodes - boolean schema (false) 1`] = `
[
  <strong>
    type
  </strong>,
  "Â :Â ",
  <strong>
    none
  </strong>,
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    Always invalid
  </div>,
]
`;

exports[`JSONSchemaViewer - corner cases CreateNodes - boolean schema (true) 1`] = `
[
  <strong>
    type
  </strong>,
  "Â :Â ",
  <strong>
    any
  </strong>,
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    Always valid
  </div>,
]
`;

exports[`JSONSchemaViewer - corner cases CreateObject - missing pattern in propertyNames 1`] = `
[
  <strong>
    type
  </strong>,
  "Â :Â ",
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    object
  </span>,
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />,
]
`;

exports[`JSONSchemaViewer - corner cases CreateObject - sub properties as boolean schema 1`] = `
[
  <strong>
    type
  </strong>,
  "Â :Â ",
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    object
  </span>,
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            property name*
          </code>
          Â 
          <strong>
            none
          </strong>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <strong>
              none
            </strong>
            <div
              style={
                {
                  "marginTop": "var(--ifm-table-cell-padding)",
                }
              }
            >
              Always invalid
            </div>
          </div>
        </div>
      </details>
    </li>
  </ul>,
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      ðŸš¨Â 
      No extra propertie(s) are authorized in this object
    </div>
  </div>,
]
`;
